hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcServer.java

```diff
@@ -67,20 +68,49 @@
    * Tests may set this down from unlimited.
    */
   public static final String HBASE_NETTY_EVENTLOOP_RPCSERVER_THREADCOUNT_KEY =
-    "hbase.netty.eventloop.rpcserver.thread.count";
+      "hbase.netty.eventloop.rpcserver.thread.count";
   private static final int EVENTLOOP_THREADCOUNT_DEFAULT = 0;
 
+  protected static final String SERVER_TCP_BACKLOG = "hbase.ipc.server.tcpbacklog";
+  protected static final String SERVER_TCP_REUSEADDR = "hbase.ipc.server.tcpreuseaddr";
+  protected static final String SERVER_TCP_NODELAY = "hbase.ipc.server.tcpnodelay";
+  protected static final String SERVER_TCP_KEEPALIVE = "hbase.ipc.server.tcpkeepalive";
+
+  protected static final String SERVER_BUFFER_LOW_WATERMARK = "hbase.ipc.server.bufferlowwatermark";
+  protected static final String SERVER_BUFFER_HIGH_WATERMARK =
+      "hbase.ipc.server.bufferhighwatermark";
+
+  protected static final boolean DEFAULT_SERVER_REUSEADDR = true;
+  protected static final int DEFAULT_SERVER_TCP_BACKLOG = 1024;
```

Comment:
Low for a server?

Reply:
Yeah, this lower end is much lower than the default.

```java
    private static final int DEFAULT_LOW_WATER_MARK = 32 * 1024;
    private static final int DEFAULT_HIGH_WATER_MARK = 64 * 1024;

    public static final WriteBufferWaterMark DEFAULT =
            new WriteBufferWaterMark(DEFAULT_LOW_WATER_MARK, DEFAULT_HIGH_WATER_MARK, false);
```

Reply:
The TCP_BACKLOG? I think the linux default is often 128 as mentioned in the JIRA? Raising it too high is also often a mistake, so I went with what the original author did. Have another suggestion?

Reply:
For the WATER_MARK settings, the previous pr is using:

 protected static final int DEFAULT_SERVER_BUFFER_LOW_WATERMARK = 1024;
 protected static final int DEFAULT_SERVER_BUFFER_HIGH_WATERMARK = 64 * 1024;

If the number of queued bytes goes above the high mark, then channel.isWritable will return false until the number of outstanding bytes falls below the low water mark.

Reply:
> The TCP_BACKLOG?

Sorry, no. I had this on the wrong line. I mean `DEFAULT_SERVER_BUFFER_LOW_WATERMARK = 1024`.

> If the number of queued bytes goes above the high mark, then channel.isWritable will return false until the number of outstanding bytes falls below the low water mark.

Right. So I'm asking you (and I guess @SteNicholas as well): why do we need to flush out the buffer down almost completely before we continue receiving data? Is this assuming that HBase clients are generally writing large amounts of data, and this to avoid thrashing back and forth between threads as we fill, flush, fill, flush? A setting this low means we'd have to drain the buffer almost entirely before returning to the user thread.

Reply:
@ndimiduk This could improve performance.